### PR TITLE
Try to use approximate placeholder regions when outputting an AscribeUserType error in borrowck

### LINKED ISSUE #114866: higher-ranked lifetime error
### Code

```Rust
fn assert_all<F, T>(_f: F)
where
    F: FnMut(&String) -> T,
{
}

fn id(x: &String) -> &String {
    x
}

fn main() {
    assert_all::<_, &String>(id);
}
```
[playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=459ad1e3e4f5951d344bfe3448119343)

### Current output

```Shell
error: higher-ranked lifetime error
  --> src/main.rs:12:5
   |
12 |     assert_all::<_, &String>(id);
   |     ^^^^^^^^^^^^^^^^^^^^^^^^
```

### Rationale and extra context

Found while minimizing #114849.  The turbofish is crucial to reproducing this error, otherwise compilation fails as seen in that issue.

I don't know enough to say what the error *should* be, but it shouldn't be this.

### PR DESCRIPTION
**Pull request by jackh726**
Fixes #114866 

Hi from GOSIM :)

### CODE REVIEW

File: tests/ui/higher-ranked/higher-ranked-lifetime-error.stderr [hunk 1]
```diff
@@ -0,0 +1,12 @@
+error[E0308]: mismatched types
+  --> $DIR/higher-ranked-lifetime-error.rs:12:5
+   |
+LL |     assert_all::<_, &String>(id);
+   |     ^^^^^^^^^^^^^^^^^^^^^^^^ one type is more general than the other
+   |
+   = note: expected reference `&String`
+              found reference `&String`
```
Comment (compiler-errors) [inline tests/ui/higher-ranked/higher-ranked-lifetime-error.stderr:8]: I'm not sure if I consider this an improvement over "higher ranked lifetime error".

File: tests/ui/async-await/return-type-notation/issue-110963-early.stderr [hunk 2]
```diff
@@ -29,9 +35,16 @@ LL | |         if !hc.check().await {
 LL | |             log_health_check_failure().await;
 LL | |         }
 LL | |     });
-   | |______^
+   | |______^ one type is more general than the other
+   |
+   = note: expected trait `Send`
+              found trait `for<'a> Send`
```
Comment (compiler-errors) [inline tests/ui/async-await/return-type-notation/issue-110963-early.stderr:41]: Why do we end up reporting this?

File: tests/ui/higher-ranked/higher-ranked-lifetime-error.stderr (hunk 1 above)
Comment (jackh726) [reply tests/ui/higher-ranked/higher-ranked-lifetime-error.stderr:8]: Yeah, the note is not all that helpful. But its actually very similar to our pre-NLL error (see output on 1.64).

I consider this more of a "get us back to where we were" than a "this is the best we can do"

Comment (jackh726) [reply tests/ui/higher-ranked/higher-ranked-lifetime-error.stderr:8]: The "one type is more general than the other" note is the most helpful part of the diagnostic, and I think that's an improvement over "higher-ranked lifetime error" alone.

File: tests/ui/async-await/return-type-notation/issue-110963-early.stderr (hunk 2 above)
Comment (jackh726) [reply tests/ui/async-await/return-type-notation/issue-110963-early.stderr:41]: I'm not sure :) seems like something with our lowering isn't quite right. I've seen something similar before, I think.

**Comment by wesleywiser** [comment]
r? @compiler-errors 

Would you mind taking another look at this? I'm not sure if your initial feedback has all been resolved as you expect or not.

**Comment by compiler-errors** [comment]
@bors r+

I still don't think I agree that this PR is better but whatever

**Pull request merged**